Python/Product/AnacondaInstallLauncher/Program.cs

```diff
@@ -0,0 +1,192 @@
+        private DateTime _stopAt;
+        private bool _cancel;
+
+        private Program(string command, string installer, string targetDir) {
+            _targetDir = Path.GetFullPath(targetDir);
+            _installer = Path.GetFullPath(installer);
+            if (!File.Exists(_installer)) {
+                Console.Error.WriteLine($"Did not find {_installer}");
+                throw new FileNotFoundException(_installer);
+            }
+            _stopAt = DateTime.UtcNow.AddMinutes(20);
+            _install = command.Equals("install", StringComparison.OrdinalIgnoreCase);
+        }
+
+        int Go() {
+            try {
+                if (_install) {
+                    return DoInstall();
+                } else {
+                    return DoUninstall();
+                }
+            } catch (Exception ex) {
+                Console.Error.WriteLine(ex);
+                Console.Error.WriteLine(ex.ToString());
+                return ex.HResult;
+            }
+        }
+
+        void Cancel() => _cancel = true;
+
+        Process Run(string filename, string arguments) {
+            var psi = new ProcessStartInfo(filename, arguments) {
+                CreateNoWindow = true,
+                RedirectStandardOutput = true,
+                RedirectStandardError = true,
+                UseShellExecute = false
+            };
+            Console.Error.WriteLine("Starting:{0}  Filename: {1}{0}  Argument: {2}", Environment.NewLine, psi.FileName, psi.Arguments);
+            var p = Process.Start(psi);
+
+            p.OutputDataReceived += (_, e) => { if (e.Data != null) Console.Write(e.Data); };
+            p.ErrorDataReceived += (_, e) => { if (e.Data != null) Console.Error.Write(e.Data); };
+
+            return p;
+        }
+
+        int DoInstall() {
+            WaitForUninstall(false);
+            if (_cancel) {
+                return ERROR_INSTALL_CANCEL;
+            }
+
+            var p = Run(_installer, string.Format(CultureInfo.InvariantCulture, "/InstallationType=AllUsers /RegisterPython=0 /S /D={0}", _targetDir));
+            p.WaitForExit();
+            return p.ExitCode;
+        }
+
+        int DoUninstall() {
+            var p = Run(_installer, "/S");
+
+            if (p.WaitForExit(30000)) {
+                Console.Error.WriteLine("Process exited: {0}", p.ExitCode);
+            } else {
+                Console.Error.WriteLine("Process has not exited");
+            }
+
+            WaitForUninstall(true);
+
+            return _cancel ? ERROR_INSTALL_CANCEL : p.ExitCode;
+        }
+
+        void WaitForUninstall(bool requireWait) {
+            try {
+                WaitForProcess(requireWait);
+            } catch (Exception ex) {
+                Console.Error.WriteLine(ex.ToString());
+            }
+        }
+
+        bool IsDirectoryPresent() {
+            if (!Directory.Exists(_targetDir)) {
+                Console.Error.WriteLine("Target directory is not present");
```

Comment:
Might be useful to print `_targetDir` at the end of all these messages

Reply:
It'll be followed up immediately by the log message for launching the installer, and it'll be in the log for starting this process too. I don't think it needs repeating much more than that.